### PR TITLE
feat(replay): add viewed-by endpoint + abstract kafka utils for replay events

### PR DESCRIPTION
**Pull request by aliu39**
Relates to https://github.com/getsentry/sentry/issues/64924

[Blueprint](https://github.com/getsentry/sentry/blob/bd8b0486473cf3136ba04169028ed76313af51ce/src/sentry/replays/blueprints/api.md?plain=1#L615)

### CODE REVIEW

File: src/sentry/replays/endpoints/organization_replay_viewed_by.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        #     fields=request.query_params.getlist("field"),
+        # )
+        #
+        # if len(response) == 0:
+        #     return Response(status=404)
+        # else:
+        #     return Response({"data": response[0]}, status=200)
+
+    # TODO:
+    # def post(self, request: Request, organization: Organization):
```
Comment (iamrajjoshi) [inline src/sentry/replays/endpoints/organization_replay_viewed_by.py:89]: Can we remove the commented code?

Comment (aliu3ntry) [reply src/sentry/replays/endpoints/organization_replay_viewed_by.py:89]: this is a draft, will mark it 😢

File: src/sentry/replays/endpoints/organization_replay_viewed_by.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                request, organization, project_ids=ALL_ACCESS_PROJECTS
+            )
+        except NoProjects:
+            return Response(status=404)
+
+        # TODO: do we need this range filter?
+        if not filter_params["start"] or not filter_params["end"]:
+            return Response(status=404)
+
+        # TODO: are there more efficient ways to lookup or pass in the project_id, rather than a snuba query?
```
Comment (JoshFerge) [inline src/sentry/replays/endpoints/organization_replay_viewed_by.py:117]: yeah, why don't we just pass in the project_id? could even make this a project scoped endpoint

File: src/sentry/replays/tasks.py [hunk 3]
```diff
@@ -104,3 +104,36 @@ def archive_replay(publisher: KafkaPublisher, project_id: int, replay_id: str) -
...
+
+
+@instrumented_task(
+    name="sentry.replays.tasks.publish_replay_viewed",
+    queue="replays.replay_viewed",
+    default_retry_delay=5,
+    max_retries=5,
+    silo_mode=SiloMode.REGION,
+)
+def publish_replay_viewed(project_id: int, replay_id: str, viewed_by_id: int) -> None:
```
Comment (JoshFerge) [inline src/sentry/replays/tasks.py:116]: do we want to put some kind of rate limiting on this endpoint per user? what if a user views a replay 100000 times with a selenium script or something?

File: src/sentry/replays/post_process.py [hunk 4]
```diff
@@ -73,13 +81,27 @@ class ReplayDetailsResponse(TypedDict, total=False):
...
+class ReplayViewedByResponse(TypedDict):
+    id: str
+    viewed_by: list[UserSerializerResponse]
 
 
 def process_raw_response(
-    response: list[dict[str, Any]], fields: list[str]
+    response: list[dict[str, Any]],
+    fields: list[str],
+    actor_uid: int | None = None,
```
Comment (JoshFerge) [inline src/sentry/replays/post_process.py:96]: why are we calling this actor_uid inster of just user_id or something?

File: src/sentry/replays/post_process.py [hunk 5]
```diff
@@ -190,6 +220,32 @@ def generate_sorted_urls(url_groups: list[tuple[int, list[str]]]) -> Iterator[st
...
 
+def generate_viewed_by_response(
+    replay_id: str,
+    viewed_by_ids: list[int],
+    actor: User,
+) -> ReplayViewedByResponse:
+    """Fetch user objects from postgres, serialize them, and format to the output expected by ReplayViewedByEndpoint."""
+    global user_serializer
+
+    # TODO: what happens if a viewed_by_id is invalid or unauthorized?
```
Comment (JoshFerge) [inline src/sentry/replays/post_process.py:231]: how could it be unauthorized?

File: src/sentry/replays/endpoints/organization_replay_viewed_by.py (hunk 2 above)
Comment (aliu3ntry) [reply src/sentry/replays/endpoints/organization_replay_viewed_by.py:117]: So move both GET and POST to a project endpoint? Makes sense to me. I was following the blueprint for this draft

File: src/sentry/replays/tasks.py (hunk 3 above)
Comment (aliu3ntry) [reply src/sentry/replays/tasks.py:116]: Mm sounds smart. How do we implement this?

File: src/sentry/replays/post_process.py (hunk 4 above)
Comment (aliu3ntry) [reply src/sentry/replays/post_process.py:96]: Was worried bout confusion with the user_id field in the replays table. Open to suggestions. I also considered:
- as_user
- request_user_id
- viewed_by_id

File: src/sentry/replays/post_process.py (hunk 5 above)
Comment (aliu3ntry) [reply src/sentry/replays/post_process.py:231]: Didn't get to the bottom of it but I suspected, from the as_user param of serialize_many, that some kind of authorization is needed for some user objects.

File: src/sentry/replays/post_process.py [hunk 6]
```diff
@@ -33,7 +41,7 @@ class BrowserResponseType(TypedDict, total=False):
 
 
 class UserResponseType(TypedDict, total=False):
-    id: str | None
+    id: str | None  # string uint64
```
Comment (cmanallen) [inline src/sentry/replays/post_process.py:44]: This id field is not a uint64.  Its a string.  It comes from the customer's application.

File: src/sentry/replays/usecases/query/__init__.py [hunk 7]
```diff
@@ -286,7 +286,12 @@ def make_full_aggregation_query(
     period_start: datetime,
     period_end: datetime,
 ) -> Query:
-    """Return a query to fetch every replay in the set."""
+    """
+    Return a query to fetch every replay in the set.
```
Comment (cmanallen) [inline src/sentry/replays/usecases/query/__init__.py:290]: Remove the line break and start the summary sentence on the first line.  Next to `"""`.

File: src/sentry/replays/endpoints/project_replay_viewed_by.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        try:
+            replay_id = str(uuid.UUID(replay_id))
+        except ValueError:
+            return Response(status=404)
+
+        # query for user ids who viewed the replay
+        filter_params = self.get_filter_params(request, project)
+        # TODO: do we need this range filter?
+        if not filter_params["start"] or not filter_params["end"]:
```
Comment (cmanallen) [inline src/sentry/replays/endpoints/project_replay_viewed_by.py:62]: This is already handled by `get_filter_params`.

File: src/sentry/replays/post_process.py [hunk 9]
```diff
@@ -190,6 +220,32 @@ def generate_sorted_urls(url_groups: list[tuple[int, list[str]]]) -> Iterator[st
...
+    replay_id: str,
+    viewed_by_ids: list[int],
+    actor: User,
+) -> ReplayViewedByResponse:
+    """Fetch user objects from postgres, serialize them, and format to the output expected by ReplayViewedByEndpoint."""
+    global user_serializer
+
+    # TODO: what happens if a viewed_by_id is invalid or unauthorized?
+
+    # method 1: hybrid cloud user_service. Based on GroupSeenSerializer. No way
```
Comment (cmanallen) [inline src/sentry/replays/post_process.py:233]: We should use the hybrid-cloud services.

**Review by cmanallen — CHANGES_REQUESTED** [review]

File: src/sentry/replays/query.py [hunk 10]
```diff
@@ -703,6 +734,11 @@ def _empty_uuids_lambda():
         parameters=[Column("count_info_events")],
         alias="count_infos",
     ),
+    "viewed_by_ids": Function(
+        "groupUniqArray",
```
Comment (cmanallen) [inline src/sentry/replays/query.py:738]: We need to remove the `0` ids.  We can ues `groupUniqArrayIf` which accepts a second argument which can be `x -> x > 0`.  Or something similar.  There's prior art in this file with `anyIf`.

**Review by cmanallen — CHANGES_REQUESTED** [review]

File: src/sentry/replays/post_process.py [hunk 11]
```diff
@@ -181,6 +203,14 @@ def generate_normalized_output(
         ret_item["info_ids"] = item.pop("info_ids", None)
         ret_item["count_infos"] = item.pop("count_infos", None)
         ret_item["count_warnings"] = item.pop("count_warnings", None)
+
+        # TODO: why are we using pop for all of these? The performance is probably worse than get
+        has_viewed = False
+        if actor_uid is not None:
+            if viewed_by_ids := item.pop("viewed_by_ids", None):
+                has_viewed = actor_uid in viewed_by_ids
+        ret_item["has_viewed"] = has_viewed
```
[the remark this answers is not in the document]
Comment (aliu3ntry) [reply src/sentry/replays/post_process.py:212]: discussed on call

**Comment by aliu3ntry** [comment]
Can I get some feedback on the `test_get_replay_viewed_by_user_fields`? Some of these fields are None or empty in the test environment. Wasn't sure how to test all of them, or whether it's necessary to.

**Comment by aliu3ntry** [comment]
Also, not sure about the use of @extend_schema_serializer() for ReplayViewedByResponse

File: src/sentry/replays/endpoints/project_replay_viewed_by.py [hunk 12]
```diff
@@ -104,8 +110,46 @@ def post(self, request: Request, project: Project, replay_id: str) -> Response:
...
         return Response(status=204)
+
+
+def _normalize_user(user: dict[str, Any]) -> dict[str, Any]:
+    """Return a normalized user dictionary.
+
+    The viewed-by resource is expected to return a subset of the user_service's
+    response output.
+    """
+    return {
```
Comment (aliu3ntry) [inline src/sentry/replays/endpoints/project_replay_viewed_by.py:125]: What are the chances the output of user_service.serialize changes, and this breaks? Should we use .get() instead?

File: src/sentry/replays/endpoints/project_replay_viewed_by.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            end=filter_params["end"],
+            request_user_id=request.user.id,
+            organization=project.organization,
+        )
+        if not viewed_by_ids_response:
+            return Response(status=404)
+
+        viewed_by_ids = viewed_by_ids_response[0]["viewed_by_ids"]
+        if viewed_by_ids == []:
+            return Response({"data": {"id": replay_id, "viewed_by": []}}, status=200)
```
Comment (aliu3ntry) [inline src/sentry/replays/endpoints/project_replay_viewed_by.py:80]: Changed the blueprint yesterday to omit the id

**Comment by cmanallen** [comment]
Sorry for the pings everyone...

**Review by cmanallen — APPROVED** [review]

**Pull request merged**